### DISCUSSION TITLE
Help to understand data.xml and layout.xml

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#490] [Q&A] [answered]
Hi, I am trying to understand this amazing software but I am struggling a lot because my background comes from a Graphic DTP experience.

Reading the manual and checking the example unfortunately is complicated cause there is not a pattern that I can recognize.

Starting very basic, and honestly that was a great personal achievement:

`data.xml`
```
<data>
  <animals title="Animals">
    <item>Dogs</item>
    <item>Cats</item>
    <item>Birds</item>
  </animals>
</data>
```

`layout.xml`
```
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Record element="data">
	  <PlaceObject>
	    <Textblock>
	      <Paragraph>
		<Value select="animals"/>
	      </Paragraph>
	    </Textblock>
	  </PlaceObject>
  </Record>
</Layout>
```

Now if I'd like to add a title in the next iteration I fail miserably, `data.xml` is the same, the layout is instead:
```
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Record element="data">
    <ForAll select="animals">
	  <PlaceObject>
	    <Textblock>
	      <Paragraph>
		<Value select="@title" />
	      </Paragraph>
	      <Paragraph>
		<Value select="animals"/>
	      </Paragraph>
	    </Textblock>
	  </PlaceObject>
    </ForAll>
  </Record>
</Layout>
```

With the latter changes is printed only the title and the item list are ignored.

It is clear that something is wrong, unfortunately it is not that much clear for me.

Can you help understand my errors and perhaps explain a little bit further...

Thanks... 🙏

**Comment by pgundlach** [comment, marked as answer]
Thank you very much for the discussion, this helps me to understand where the manual is lacking. I try to improve.

Now for your question. The key to understand the problem is the “focus” of the data processing. There is always a "current element" in the data xml file, which can change when using the `<ProcessNode>` or the `<ForAll>` command.

Let my try to explain the first layout, which works, but not as you might think:

```xml
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

	<Record element="data">
		<PlaceObject>
			<Textblock>
				<Paragraph>
					<Value select="animals" />
				</Paragraph>
			</Textblock>
		</PlaceObject>
	</Record>
</Layout>
```

When the layout processing starts at `<Record element="data">`, the current XML node is the root element, `<data>` with all its contents. Now the `<Value select="animals" />` selects the `animals` element and its child nodes:

```xml
<animals title="Animals">
  <item>Dogs</item>
  <item>Cats</item>
  <item>Birds</item>
</animals>
```

The string value (which is what the `<Paragraph>` command wants) is Dogs Cats Birds and spaces and newlines which result from the space between the angle brackets:

<img width="308" alt="whitespace" src="https://user-images.githubusercontent.com/209434/232991155-d1e03c46-135a-47fa-a3a9-c1ff3e10573f.png">

So the output is 

```
Dogs
Cats
Birds
```

which includes the newlines you get by selecting the animals element.

Now for the second layout:


```xml
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

	<Record element="data">
		<ForAll select="animals">
			<PlaceObject>
				<Textblock>
					<Paragraph>
						<Value select="@title" />
					</Paragraph>
					<Paragraph>
						<Value select="animals" />
					</Paragraph>
				</Textblock>
			</PlaceObject>
		</ForAll>
	</Record>
</Layout>
```

in line 5 the current element is the root element `data` (as above). Now with `<ForAll>`, you change the current element to each of the child elements named `animals`. In the data file there is only one child named `animals`, so the contents of `<ForAll>` only gets executed once. The `<ForAll>` changes the current element to `animals` in each iteration. 

Now that you are at the `animals` element, you look at the attribute title (`@title`) which happens to be an attribute of the current element (`animals`). But the second `Value` command looks at all children named `animals` of the current element, but there is none, only elements called `item`. So the result of the second `Value` command is the empty sequence which is “nothing”.

Since I don't know what you really want to achieve, I can just guess what you should do:

```xml
<Layout
	xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en">

	<Record element="data">
		<ProcessNode select="animals" />
	</Record>

	<Record element="animals">
		<PlaceObject>
			<Textblock>
				<Paragraph>
					<Value select="@title" />
				</Paragraph>
				<ForAll select="item">
					<Paragraph>
						<Value select="." />
					</Paragraph>
				</ForAll>
			</Textblock>
		</PlaceObject>
	</Record>
</Layout>
```

This would instruct the speedata Publisher to go to the `<Record element="animals">` command when it sees an element called `animals` as the child node of `data`. Now the current element is `animals` as above and you can output the title and use `<ForAll>` to select  each of the children and create a paragraph for each. The `<ForAll>` sets the current element to each `item` element, so you can access the string value of the current element with `.`.  The `<ForAll>` gets executed three times, as there are three child elements named item.

**Reply by Cicorione** [reply]
Thank you very much... I am crying... 😭

I can't guarantee I understood 100% what you wrote above, but this has been amazing nonetheless!

I'll try doing my best to learn!

**Reply by pgundlach** [reply]
If you are still interested, don't hesitate to ask more questions! I try my very best (but sometimes I am a bit slow to respond).

**Reply by Cicorione** [reply]
I am totally in, but very poor oriented to think in abstraction... 😖
